### PR TITLE
ModernBERT FlexAttention

### PR DESCRIPTION
**Pull request by staghado**
# What does this PR do?
This PR adds FlexAttention support for ModernBERT:
- Combines sliding window and document masking to implement the alternating local/global attention pattern in ModernBERT
- Mask creation is expensive so the two masks are cached at the model level and then re-used across layers.
- Similar to the FA2 path, it works directly on the unpadded sequences
- Re-uses the existing ModernBertRotaryEmbedding to avoid requiring FA2.

Note:
The current version requires one of the latest torch nightlies (e.g 2.6.0.dev20241112)
Currently transformers does not allow compiling the flex_attention function IIUC

### CODE REVIEW

**Comment by tomaarsen** [comment]
I think this is very interesting, but only if the performance rivals that of e.g. SDPA. I see your issue here: https://github.com/pytorch-labs/attention-gym/issues/95, which also shows that with compilation, Flex Attention outperforms SDPA in a lot of common cases, so we would have to introduce the compilation option for Flex Attention in `transformers` before this is viable I think.

I'd love to see that implemented, though!

- Tom Aarsen

**Comment by staghado** [comment]
Yeah when compiled FlexAttention is generally faster than SDPA and has much lower memory from my tests.
How would we go about adding the compilation option? would a flag suffice in this case?

**Comment by staghado** [comment]
Any ideas on how transformers plans to support FlexAttention compilation? I see that no models have it implemented for now(I might be wrong here)
cc @tomaarsen @ArthurZucker

**Review by ArthurZucker — COMMENTED** [review]
Looks alright appart from the nightly requirement! 
Regarding compilation, we do support compile, and by default when you use generate with `cache_implementation="static"`.

**Comment by staghado** [comment]
I think the nightly requirement might be resolved by specifying the `BLOCK_SIZE` argument in `create_block_mask`, will verify that.
for compilation, I meant supporting the `flex_attention = torch.compile(flex_attention, dynamic=False)` which is very important for performance as mentioned [here](https://github.com/pytorch-labs/attention-gym/issues/95).
It also looks like the flexattention in Gemma was removed?

**Comment by neavo** [comment]
> # What does this PR do?
> This PR adds FlexAttention support for ModernBERT:
> 
> * Combines sliding window and document masking to implement the alternating local/global attention pattern in ModernBERT
> * Mask creation is expensive so the two masks are cached at the model level and then re-used across layers.
> * Similar to the FA2 path, it works directly on the unpadded sequences
> * Re-uses the existing ModernBertRotaryEmbedding to avoid requiring FA2.
> 
> Note: The current version requires one of the latest torch nightlies (e.g 2.6.0.dev20241112) Currently transformers does not allow compiling the flex_attention function IIUC

Torch 2.6 has released the RC version. The dependency issues should not be a big problem anymore. Looking forward to its merging.
Before that, could you please update the branch of this PR to make its code compatible with the current mainline version?
I would like to give it a try.

File: src/transformers/models/modernbert/modeling_modernbert.py [hunk 1]
```diff
@@ -51,6 +51,14 @@
 else:
     RotaryEmbedding = object
 
+if is_torch_flex_attn_available():
+    from torch.nn.attention.flex_attention import BlockMask, create_block_mask, flex_attention
+
+    flex_attention = torch.compile(flex_attention)
```
Comment (staghado) [inline src/transformers/models/modernbert/modeling_modernbert.py:57]: these are needed to actually use the flexattention kernels but the `utils/modular_model_converter.py` does not allow it in the converted file. 
let me know if there is a better way to do this

Comment (staghado) [reply src/transformers/models/modernbert/modeling_modernbert.py:57]: cc @ArthurZucker @tomaarsen (sorry for the ping)

**Comment by staghado** [comment]
With the release of [PyTorch 2.6](https://pytorch.org/blog/pytorch2-6/), it is now possible to use FlexAttention with ModernBERT.

File: src/transformers/models/modernbert/modeling_modernbert.py (hunk 1 above)
Comment (staghado) [reply src/transformers/models/modernbert/modeling_modernbert.py:57]: any ideas on how to support compiling the flex_attention function in transformers?

Comment (staghado) [reply src/transformers/models/modernbert/modeling_modernbert.py:57]: The failing CI is related to these two lines, there is no clear way how to support compile in a clean way

Comment (ArthurZucker) [reply src/transformers/models/modernbert/modeling_modernbert.py:57]: will have a look in a bit sorry about that!

**Review by ArthurZucker — COMMENTED** [review]
Okay! Sorry for the delay, pieces of the puzzle are coming one at a time 🤣 
#36103 should add something similar so we can take inspiration on it IMO.
If possible we want the least amount of changes in the modernbert modeling file, but otherwise sound great! let's follow a bit what is done in the linked PR, potentially using the same function to create the block mask!

**Comment by ArthurZucker** [comment]
Sorry @staghado for the delay!

**Comment by staghado** [comment]
I skimmed through the linked PR, flex_attention does not get compiled automatically so the compilation call is needed. As you can see [here](https://github.com/pytorch-labs/attention-gym/issues/95).
Another thing is the integration uses score_mod which is less performant than block mask as mentioned [here](https://pytorch.org/blog/flexattention/#q-how-can-we-compute-blockmask-quicker)

I would prefer keeping the ModernBERT implementation separate for now, at least until the integration is more mature.

**Comment by bursteratom** [comment]
Hi @staghado my PR https://github.com/huggingface/transformers/pull/36103/files does add torch compile for flex attention, it is done in the `integrations/flex_attention.py` such that minimal modifications are required in the various `modeling_{model type}.py` for them to take advantage of flex attention.

**Comment by ArthurZucker** [comment]
Hey #36643 was merged, just needs to integrate the refactor points imo and I'll have another look!

**Comment by staghado** [comment]
Looks like the integration is too focused on decoder-only models(gqa=true at api level), create_block_mask is renamed to create_block_causal_mask_flex, a score_mod is always passed even if not needed. This PR is out of scope.
